pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1121 @@
+			if let Some(who) = account {
+				T::Currency::transfer(asset, &who, &Self::protocol_account(), amount)?;
+
+				// if provided by LP, create and mint a position instance
+				let lp_position = Position::<T::Balance, T::AssetId> {
+					asset_id: asset,
+					amount,
+					shares: amount,
+					price: Position::<T::Balance, T::AssetId>::price_to_balance(initial_price),
+				};
+
+				let instance_id = Self::create_and_mint_position_instance(&who)?;
+
+				<Positions<T>>::insert(instance_id, lp_position);
+
+				Self::deposit_event(Event::PositionCreated {
+					position_id: instance_id,
+					owner: who,
+					asset,
+					amount,
+					shares: amount,
+					price: initial_price,
+				});
+			} else {
+				// Ensure that it has been transferred to protocol account by other means
+				ensure!(
+					T::Currency::free_balance(asset, &Self::protocol_account()) >= amount,
+					Error::<T>::MissingBalance
+				);
+			}
+
+			// Imbalance update
+			Self::recalculate_imbalance(&state, BalanceUpdate::Decrease(amount))?;
+
+			// Total hub asset liquidity update
+			Self::update_hub_asset_liquidity(
+				&BalanceUpdate::Increase(hub_reserve),
+				HubAssetIssuanceUpdate::AdjustSupply,
+			)?;
+
+			// TVL update
+			if stable_asset_reserve != T::Balance::zero() && stable_asset_hub_reserve != T::Balance::zero() {
+				<TotalTVL<T>>::try_mutate(|tvl| -> DispatchResult {
+					*tvl = initial_price
+						.checked_mul(&Price::from((stable_asset_reserve, stable_asset_hub_reserve)))
+						.and_then(|v| v.checked_mul_int(amount))
+						.and_then(|v| v.checked_add(&*tvl))
+						.ok_or(Error::<T>::Overflow)?;
+					Ok(())
+				})?;
+			}
+
+			<Assets<T>>::insert(asset, state);
+
+			Self::deposit_event(Event::TokenAdded {
+				asset_id: asset,
+				initial_amount: amount,
+				initial_price,
+			});
+
+			Ok(())
+		}
+
+		/// Add liquidity of asset `asset` in quantity `amount` to Omnipool
+		///
+		/// add_liquidity adds specified asset amount to pool and in exchange gives the origin
+		/// corresponding shares amount in form of NFT at current price.
+		///
+		/// NFT is minted using NTFHandler which implements non-fungibles traits from frame_support.
+		///
+		/// Parameters:
+		/// - `asset`: The identifier of the new asset added to the pool. Must be already in the pool
+		/// - `amount`: Amount of asset added to omnipool
+		///
+		/// Emits `LiquidityAdded` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(origin: OriginFor<T>, asset: T::AssetId, amount: T::Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinimumPoolLiquidity::get(),
+				Error::<T>::InsufficientLiquidity
+			);
+
+			ensure!(
+				T::Currency::free_balance(asset, &who) >= amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let mut asset_state = Assets::<T>::get(asset).ok_or(Error::<T>::AssetNotFound)?;
```

Comment:
refactored to use try_mutate. 

however, it does not change much right now, because the scope is basically the same, only in the try_mutate. 
this is due to the fact, that mutable reference is still needed at the very end when TVL is updated. 

tvl updates will be refactored ( made part of the math) and then, the scope of asset state can be reduced as mutable reference might be no longer needed.